### PR TITLE
Introduce database name property for Postgres database resources

### LINKED ISSUE #1199: The argument for AddDatabase represents a database and a connection string name at the same time
Let's take this example for PostgreSQL:

```csharp
var catalogDb = builder.AddPostgresContainer("postgres").AddDatabase("catalogdb");
```

image

`"catalogdb"` is both the name of the database (a technical name), and the name of the resource (a logical name).
 
As a developer when I read `AddDatabase(string name)` I understand that is can be a database name, as in `CREATE DATABASE "catalogdb"`.
 
But in the app when I do `builder.AddNpgsqlDbContext<CatalogDbContext>("catalogdb")` it's asking for a `connectionName`.
 
Though the documentation for AddDatabase(string name) is clear:

> The name of the resource. This name will be used as the connection string name when referenced in a dependency.
 
It seems like we should be able to provide a database name and a resource name (or connection string name), because we might want them to be different, or change independently.

The `name` argument should also be changed to remove the ambiguity.

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #1199

Only for Postgres for the moment to validate the solution.

Open questions
- We should apply this to all databases, any other resources we need for sure? @mitchdenny was saying anything with a child resource potentially.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2346)

### CODE REVIEW

**Comment by davidfowl** [comment]
Yea do it for all of the databases.

File: tests/Aspire.Hosting.Tests/SqlServer/AddSqlServerTests.cs [hunk 1]
```diff
@@ -134,4 +134,67 @@ public void VerifyManifest()
...
+    [Fact]
+    public void ThrowsWithIdenticalChildResourceNames()
+    {
+        var builder = DistributedApplication.CreateBuilder();
+
+        builder.AddSqlServer("sqlserver1")
+            .AddDatabase("db");
+
+        Assert.Throws<DistributedApplicationException>(() =>
+            builder.AddSqlServer("sqlserver1")
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply tests/Aspire.Hosting.Tests/SqlServer/AddSqlServerTests.cs]: Done

File: src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs [hunk 2]
```diff
@@ -36,11 +36,16 @@ public static IResourceBuilder<MongoDBServerResource> AddMongoDB(this IDistribut
     /// Adds a MongoDB database to the application model.
     /// </summary>
     /// <param name="builder">The MongoDB server resource builder.</param>
-    /// <param name="name">The name of the resource. This name will be used as the connection string name when referenced in a dependency.</param>
+    /// <param name="name">The name of the resource. If <paramref name="databaseName"/> is not provided this name will be used as the connection string name when referenced in a dependency and must be unique.</param>
+    /// <param name="databaseName">The name of the database if different from <paramref name="name"/>.</param>
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs]: Done

File: src/Aspire.Hosting/MySql/MySqlServerResource.cs [hunk 3]
```diff
@@ -41,6 +41,23 @@ public class MySqlServerResource(string name, string password) : ContainerResour
...
+    public IReadOnlyDictionary<string, string> Databases => _databases;
+
+    internal void AddDatabase(string name, string databaseName)
+    {
+        if (_databases.ContainsKey(name))
+        {
+            return;
+        }
+
+        _databases[name] = databaseName;
```
Comment (eerhardt) [inline src/Aspire.Hosting/MySql/MySqlServerResource.cs:58]:
```suggestion
        _databases.TryAdd(name, databaseName);
```

File: src/Aspire.Hosting/Oracle/OracleDatabaseServerResource.cs [hunk 4]
```diff
@@ -41,6 +41,23 @@ public class OracleDatabaseServerResource(string name, string password) : Contai
...
+    public IReadOnlyDictionary<string, string> Databases => _databases;
+
+    internal void AddDatabase(string name, string databaseName)
+    {
+        if (_databases.ContainsKey(name))
+        {
+            return;
+        }
+
+        _databases[name] = databaseName;
```
Comment (eerhardt) [inline src/Aspire.Hosting/Oracle/OracleDatabaseServerResource.cs:58]:
```suggestion
        _databases.TryAdd(name, databaseName);
```

File: src/Aspire.Hosting/Postgres/PostgresServerResource.cs [hunk 5]
```diff
@@ -57,21 +56,21 @@ public class PostgresServerResource(string name, string password) : ContainerRes
...
+    /// A dictionary where the key is the resource name and the value is the database name.
     /// </summary>
-    public IEnumerable<string> Databases => _databases.ToImmutableArray();
+    public IReadOnlyDictionary<string, string> Databases => _databases;
 
-    internal void AddDatabase(string databaseName)
+    internal void AddDatabase(string name, string databaseName)
     {
-        if (_databases.Contains(databaseName, StringComparers.ResourceName))
+        if (_databases.ContainsKey(name))
```
Comment (eerhardt) [inline src/Aspire.Hosting/Postgres/PostgresServerResource.cs:68]: Same here `_databases.TryAdd(name, databaseName);`

File: src/Aspire.Hosting/SqlServer/SqlServerServerResource.cs [hunk 6]
```diff
@@ -58,21 +57,21 @@ public class SqlServerServerResource(string name, string password) : ContainerRe
...
+    /// A dictionary where the key is the resource name and the value is the database name.
     /// </summary>
-    public IEnumerable<string> Databases => _databases.ToImmutableArray();
+    public IReadOnlyDictionary<string, string> Databases => _databases;
 
-    internal void AddDatabase(string databaseName)
+    internal void AddDatabase(string name, string databaseName)
     {
-        if (_databases.Contains(databaseName, StringComparers.ResourceName))
+        if (_databases.ContainsKey(name))
```
Comment (eerhardt) [inline src/Aspire.Hosting/SqlServer/SqlServerServerResource.cs:69]: Same here `_databases.TryAdd(name, databaseName);`

**Review by eerhardt — APPROVED** [review]

File: src/Aspire.Hosting/MySql/MySqlServerResource.cs (hunk 3 above)
Comment (sebastienros) [reply src/Aspire.Hosting/MySql/MySqlServerResource.cs:58]: I knew it, then I forgot 🤦

**Pull request merged**